From c3e7caeb969771e1364c3eeab91d8adb66e9ea01 Mon Sep 17 00:00:00 2001
From: Li Wei <liwei825@huawei.com>
Date: Wed, 15 Feb 2023 16:36:05 +0800
Subject: [PATCH 320/921] [Huawei] ocfs2/cluster: add the customized
 modification of nodemanager

Offering: HULK
ocfs2 inclusion
category: feature
bugzilla: NA

----------------------------------------

Add the customized modification of nodemanager and related files.

Signed-off-by: Li Wei <liwei825@huawei.com>
---
 fs/ocfs2/cluster/Makefile            |   1 +
 fs/ocfs2/cluster/masklog.c           |  11 +-
 fs/ocfs2/cluster/masklog.h           |  22 +-
 fs/ocfs2/cluster/nodemanager.c       | 444 ++++++++++++++++++++++++---
 fs/ocfs2/cluster/nodemanager.h       |  30 +-
 fs/ocfs2/cluster/ocfs2_nodemanager.h |   4 +-
 6 files changed, 447 insertions(+), 65 deletions(-)

diff --git a/fs/ocfs2/cluster/Makefile b/fs/ocfs2/cluster/Makefile
index 0e5b73215819..2af42f690dd6 100644
--- a/fs/ocfs2/cluster/Makefile
+++ b/fs/ocfs2/cluster/Makefile
@@ -1,4 +1,5 @@
 # SPDX-License-Identifier: GPL-2.0-only
+ccflags-y := -DCONFIG_OCFS2_DEBUG_MASKLOG
 obj-$(CONFIG_OCFS2_FS) += ocfs2_nodemanager.o
 
 ocfs2_nodemanager-objs := heartbeat.o masklog.o sys.o nodemanager.o \
diff --git a/fs/ocfs2/cluster/masklog.c b/fs/ocfs2/cluster/masklog.c
index 1d696c96b8b2..e201055719b8 100644
--- a/fs/ocfs2/cluster/masklog.c
+++ b/fs/ocfs2/cluster/masklog.c
@@ -16,7 +16,7 @@
 
 struct mlog_bits mlog_and_bits = MLOG_BITS_RHS(MLOG_INITIAL_AND_MASK);
 EXPORT_SYMBOL_GPL(mlog_and_bits);
-struct mlog_bits mlog_not_bits = MLOG_BITS_RHS(0);
+struct mlog_bits mlog_not_bits = MLOG_BITS_RHS(MLOG_INITIAL_NOT_MASK);
 EXPORT_SYMBOL_GPL(mlog_not_bits);
 
 static ssize_t mlog_mask_show(u64 mask, char *buf)
@@ -105,21 +105,28 @@ static struct mlog_attribute mlog_attrs[MLOG_MAX_BITS] = {
 	define_mask(SOCKET),
 	define_mask(HEARTBEAT),
 	define_mask(HB_BIO),
-	define_mask(DLMFS),
+	define_mask(LOCKFS),
 	define_mask(DLM),
 	define_mask(DLM_DOMAIN),
 	define_mask(DLM_THREAD),
 	define_mask(DLM_MASTER),
 	define_mask(DLM_RECOVERY),
 	define_mask(DLM_GLUE),
+	define_mask(ADL_GLUE),
+	define_mask(LOCK_GLUE),
+	define_mask(ADL),
+	define_mask(ADL_DOMAIN),
 	define_mask(VOTE),
 	define_mask(CONN),
 	define_mask(QUORUM),
 	define_mask(BASTS),
 	define_mask(CLUSTER),
+	define_mask(FILE_IO),
+	define_mask(RO),
 	define_mask(ERROR),
 	define_mask(NOTICE),
 	define_mask(KTHREAD),
+	define_mask(DEBUG),
 };
 
 static struct attribute *mlog_attr_ptrs[MLOG_MAX_BITS] = {NULL, };
diff --git a/fs/ocfs2/cluster/masklog.h b/fs/ocfs2/cluster/masklog.h
index 446e452ac7a6..3270ee914514 100644
--- a/fs/ocfs2/cluster/masklog.h
+++ b/fs/ocfs2/cluster/masklog.h
@@ -73,7 +73,7 @@
 #define ML_SOCKET	0x0000000000000004ULL /* net socket lifetime */
 #define ML_HEARTBEAT	0x0000000000000008ULL /* hb all heartbeat tracking */
 #define ML_HB_BIO	0x0000000000000010ULL /* hb io tracing */
-#define ML_DLMFS	0x0000000000000020ULL /* dlm user dlmfs */
+#define ML_LOCKFS	0x0000000000000020ULL /* lock user lockfs */
 #define ML_DLM		0x0000000000000040ULL /* dlm general debugging */
 #define ML_DLM_DOMAIN	0x0000000000000080ULL /* dlm domain debugging */
 #define ML_DLM_THREAD	0x0000000000000100ULL /* dlm domain thread */
@@ -85,13 +85,22 @@
 #define ML_QUORUM	0x0000000000004000ULL /* net connection quorum */
 #define ML_BASTS	0x0000000000008000ULL /* dlmglue asts and basts */
 #define ML_CLUSTER	0x0000000000010000ULL /* cluster stack */
+#define ML_FILE_IO	0x0000000000020000ULL /* ocfs2 file I/O */
+#define ML_ADL_GLUE	0x0000000000040000ULL /* ocfs2 adl glue layer */
+#define ML_ADL		0x0000000000080000ULL /* ocfs2 adl internal */
+#define ML_LOCK_GLUE	0x0000000000100000ULL /* ocfs2 lock glue layer */
+#define ML_ADL_DOMAIN	0x0000000000200000ULL
+#define ML_RO		0x0000000000400000ULL
 
 /* bits that are infrequently given and frequently matched in the high word */
 #define ML_ERROR	0x1000000000000000ULL /* sent to KERN_ERR */
 #define ML_NOTICE	0x2000000000000000ULL /* setn to KERN_NOTICE */
 #define ML_KTHREAD	0x4000000000000000ULL /* kernel thread activity */
+#define ML_DEBUG	0x8000000000000000ULL /* sent to KERN_DEBUG */
 
-#define MLOG_INITIAL_AND_MASK (ML_ERROR|ML_NOTICE)
+#define MLOG_INITIAL_AND_MASK (ML_ERROR|ML_NOTICE|ML_DLM_RECOVERY|		\
+		ML_QUORUM|ML_FILE_IO)
+#define MLOG_INITIAL_NOT_MASK (ML_DEBUG)
 #ifndef MLOG_MASK_PREFIX
 #define MLOG_MASK_PREFIX 0
 #endif
@@ -173,7 +182,7 @@ do {									\
 		mlog(mask, fmt, ##__VA_ARGS__);				\
 } while (0)
 
-#define mlog_errno(st) ({						\
+#define mlog_errno(st)({						\
 	int _st = (st);							\
 	if (_st != -ERESTARTSYS && _st != -EINTR &&			\
 	    _st != AOP_TRUNCATED_PAGE && _st != -ENOSPC &&		\
@@ -182,6 +191,13 @@ do {									\
 	_st;								\
 })
 
+#define mlog_errno_debug(st)({						\
+	int _st = (st);							\
+	if (_st != -ERESTARTSYS && _st != -EINTR &&			\
+	    _st != AOP_TRUNCATED_PAGE && _st != -EIOCBQUEUED)		\
+		mlog(ML_DEBUG, "status = %lld\n", (long long)_st);	\
+})
+
 #define mlog_bug_on_msg(cond, fmt, args...) do {			\
 	if (cond) {							\
 		mlog(ML_ERROR, "bug expression: " #cond "\n");		\
diff --git a/fs/ocfs2/cluster/nodemanager.c b/fs/ocfs2/cluster/nodemanager.c
index 7a7640c59f3c..d087bc3c27df 100644
--- a/fs/ocfs2/cluster/nodemanager.c
+++ b/fs/ocfs2/cluster/nodemanager.c
@@ -9,6 +9,7 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/configfs.h>
+#include <linux/version.h>
 
 #include "tcp.h"
 #include "nodemanager.h"
@@ -21,15 +22,16 @@
  * cluster references throughout where nodes are looked up */
 struct o2nm_cluster *o2nm_single_cluster = NULL;
 
-static const char *o2nm_fence_method_desc[O2NM_FENCE_METHODS] = {
+static char *o2nm_fence_method_desc[O2NM_FENCE_METHODS] = {
 	"reset",	/* O2NM_FENCE_RESET */
 	"panic",	/* O2NM_FENCE_PANIC */
+	"invalid",	/* O2NM_FENCE_INVALID */
 };
 
 static inline void o2nm_lock_subsystem(void);
 static inline void o2nm_unlock_subsystem(void);
 
-struct o2nm_node *o2nm_get_node_by_num(u8 node_num)
+struct o2nm_node *o2nm_get_node_by_num(u16 node_num)
 {
 	struct o2nm_node *node = NULL;
 
@@ -63,12 +65,12 @@ int o2nm_configured_node_map(unsigned long *map, unsigned bytes)
 }
 EXPORT_SYMBOL_GPL(o2nm_configured_node_map);
 
-static struct o2nm_node *o2nm_node_ip_tree_lookup(struct o2nm_cluster *cluster,
+static struct o2nm_node *o2nm_node_ipv4_tree_lookup(struct o2nm_cluster *cluster,
 						  __be32 ip_needle,
 						  struct rb_node ***ret_p,
 						  struct rb_node **ret_parent)
 {
-	struct rb_node **p = &cluster->cl_node_ip_tree.rb_node;
+	struct rb_node **p = &cluster->cl_node_ipv4_tree.rb_node;
 	struct rb_node *parent = NULL;
 	struct o2nm_node *node, *ret = NULL;
 
@@ -98,7 +100,61 @@ static struct o2nm_node *o2nm_node_ip_tree_lookup(struct o2nm_cluster *cluster,
 	return ret;
 }
 
-struct o2nm_node *o2nm_get_node_by_ip(__be32 addr)
+static struct o2nm_node *o2nm_node_ipv6_tree_lookup(struct o2nm_cluster *cluster,
+						  u8 ip_needle[],
+						  struct rb_node ***ret_p,
+						  struct rb_node **ret_parent)
+{
+	struct rb_node **p = &cluster->cl_node_ipv6_tree.rb_node;
+	struct rb_node *parent = NULL;
+	struct o2nm_node *node, *ret = NULL;
+
+	while (*p) {
+		int cmp;
+
+		parent = *p;
+		node = rb_entry(parent, struct o2nm_node, nd_ip_node);
+
+		cmp = memcmp(ip_needle, node->nd_ipv6_address,
+				sizeof(node->nd_ipv6_address));
+		if (cmp < 0)
+			p = &(*p)->rb_left;
+		else if (cmp > 0)
+			p = &(*p)->rb_right;
+		else {
+			ret = node;
+			break;
+		}
+	}
+
+	if (ret_p != NULL)
+		*ret_p = p;
+	if (ret_parent != NULL)
+		*ret_parent = parent;
+
+	return ret;
+}
+
+struct o2nm_node *o2nm_get_node_by_ipv4(__be32 addr)
+{
+	struct o2nm_node *node = NULL;
+	struct o2nm_cluster *cluster = o2nm_single_cluster;
+
+	if (cluster == NULL)
+		goto out;
+
+	read_lock(&cluster->cl_nodes_lock);
+	node = o2nm_node_ipv4_tree_lookup(cluster, addr, NULL, NULL);
+	if (node)
+		config_item_get(&node->nd_item);
+	read_unlock(&cluster->cl_nodes_lock);
+
+out:
+	return node;
+}
+EXPORT_SYMBOL_GPL(o2nm_get_node_by_ipv4);
+
+struct o2nm_node *o2nm_get_node_by_ipv6(u8 addr[])
 {
 	struct o2nm_node *node = NULL;
 	struct o2nm_cluster *cluster = o2nm_single_cluster;
@@ -107,7 +163,7 @@ struct o2nm_node *o2nm_get_node_by_ip(__be32 addr)
 		goto out;
 
 	read_lock(&cluster->cl_nodes_lock);
-	node = o2nm_node_ip_tree_lookup(cluster, addr, NULL, NULL);
+	node = o2nm_node_ipv6_tree_lookup(cluster, addr, NULL, NULL);
 	if (node)
 		config_item_get(&node->nd_item);
 	read_unlock(&cluster->cl_nodes_lock);
@@ -115,7 +171,7 @@ struct o2nm_node *o2nm_get_node_by_ip(__be32 addr)
 out:
 	return node;
 }
-EXPORT_SYMBOL_GPL(o2nm_get_node_by_ip);
+EXPORT_SYMBOL_GPL(o2nm_get_node_by_ipv6);
 
 void o2nm_node_put(struct o2nm_node *node)
 {
@@ -129,9 +185,9 @@ void o2nm_node_get(struct o2nm_node *node)
 }
 EXPORT_SYMBOL_GPL(o2nm_node_get);
 
-u8 o2nm_this_node(void)
+u16 o2nm_this_node(void)
 {
-	u8 node_num = O2NM_MAX_NODES;
+	u16 node_num = O2NM_MAX_NODES;
 
 	if (o2nm_single_cluster && o2nm_single_cluster->cl_has_local)
 		node_num = o2nm_single_cluster->cl_local_node;
@@ -140,6 +196,15 @@ u8 o2nm_this_node(void)
 }
 EXPORT_SYMBOL_GPL(o2nm_this_node);
 
+int o2nm_fence_method_is_invalid(void)
+{
+	if (o2nm_single_cluster)
+		return (o2nm_single_cluster->cl_fence_method == O2NM_FENCE_INVALID);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(o2nm_fence_method_is_invalid);
+
 /* node configfs bits */
 
 static struct o2nm_cluster *to_o2nm_cluster(struct config_item *item)
@@ -180,6 +245,7 @@ enum {
 	O2NM_NODE_ATTR_NUM = 0,
 	O2NM_NODE_ATTR_PORT,
 	O2NM_NODE_ATTR_ADDRESS,
+	O2NM_NODE_ATTR_IPV6_ADDRESS,
 };
 
 static ssize_t o2nm_node_num_store(struct config_item *item, const char *page,
@@ -202,9 +268,14 @@ static ssize_t o2nm_node_num_store(struct config_item *item, const char *page,
 	 * node number and try to use our address and port attributes
 	 * to connect to this node.. make sure that they've been set
 	 * before writing the node attribute? */
-	if (!test_bit(O2NM_NODE_ATTR_ADDRESS, &node->nd_set_attributes) ||
-	    !test_bit(O2NM_NODE_ATTR_PORT, &node->nd_set_attributes))
-		return -EINVAL; /* XXX */
+	if (!test_bit(O2NM_NODE_ATTR_PORT, &node->nd_set_attributes))
+		return -EINVAL;
+	if ((node->nd_ipnet_type == IPV4_TYPE) &&
+	    !test_bit(O2NM_NODE_ATTR_ADDRESS, &node->nd_set_attributes))
+		return -EINVAL;
+	else if ((node->nd_ipnet_type == IPV6_TYPE) &&
+	    !test_bit(O2NM_NODE_ATTR_IPV6_ADDRESS, &node->nd_set_attributes))
+		return -EINVAL;
 
 	o2nm_lock_subsystem();
 	cluster = to_o2nm_cluster_from_node(node);
@@ -232,12 +303,12 @@ static ssize_t o2nm_node_num_store(struct config_item *item, const char *page,
 
 	return count;
 }
-static ssize_t o2nm_node_ipv4_port_show(struct config_item *item, char *page)
+static ssize_t o2nm_node_ip_port_show(struct config_item *item, char *page)
 {
-	return sprintf(page, "%u\n", ntohs(to_o2nm_node(item)->nd_ipv4_port));
+	return sprintf(page, "%u\n", ntohs(to_o2nm_node(item)->nd_ip_port));
 }
 
-static ssize_t o2nm_node_ipv4_port_store(struct config_item *item,
+static ssize_t o2nm_node_ip_port_store(struct config_item *item,
 					 const char *page, size_t count)
 {
 	struct o2nm_node *node = to_o2nm_node(item);
@@ -255,26 +326,84 @@ static ssize_t o2nm_node_ipv4_port_store(struct config_item *item,
 
 	if (test_and_set_bit(O2NM_NODE_ATTR_PORT, &node->nd_set_attributes))
 		return -EBUSY;
-	node->nd_ipv4_port = htons(tmp);
+	node->nd_ip_port = htons(tmp);
 
 	return count;
 }
 
+static int o2nm_node_change_ip_address(struct o2nm_node *node)
+{
+	int ret = 0;
+
+	/*
+	 * OTHER NODES: disconnect --> reconnect
+	 * MY NODE: stop listening --> start listening --> connect all nodes
+	 */
+	if (node->nd_num != o2nm_this_node()) {
+		o2net_disconnect_node(node);
+		o2net_connect_node(node);
+	} else {
+		o2net_stop_listening(node, true);
+		o2net_pre_stop_listening(node, true);
+		ret = o2net_start_listening(node, true);
+		if (ret) {
+			mlog_errno(ret);
+			goto out;
+		}
+
+		ret = o2net_pre_start_listening(node, true);
+		if (ret) {
+			mlog_errno(ret);
+			o2net_stop_listening(node, true);
+		}
+
+		o2net_connect_all_node(node);
+	}
+
+out:
+	return ret;
+}
+
 static ssize_t o2nm_node_ipv4_address_show(struct config_item *item, char *page)
 {
 	return sprintf(page, "%pI4\n", &to_o2nm_node(item)->nd_ipv4_address);
 }
 
+static void o2nm_node_ip_tree_rollback(struct o2nm_cluster *cluster,
+					    struct o2nm_node *node)
+{
+	struct rb_node **p, *parent;
+
+	o2nm_node_change_ip_address(node);
+
+	write_lock(&cluster->cl_nodes_lock);
+	if (node->nd_ipnet_type == IPV4_TYPE) {
+		if (!o2nm_node_ipv4_tree_lookup(cluster, node->nd_ipv4_address, &p, &parent)) {
+			rb_link_node(&node->nd_ip_node, parent, p);
+			rb_insert_color(&node->nd_ip_node, &cluster->cl_node_ipv4_tree);
+		}
+	} else if (node->nd_ipnet_type == IPV6_TYPE) {
+		if (!o2nm_node_ipv6_tree_lookup(cluster, node->nd_ipv6_address, &p, &parent)) {
+			rb_link_node(&node->nd_ip_node, parent, p);
+			rb_insert_color(&node->nd_ip_node, &cluster->cl_node_ipv6_tree);
+		}
+	}
+	write_unlock(&cluster->cl_nodes_lock);
+}
+
 static ssize_t o2nm_node_ipv4_address_store(struct config_item *item,
 					    const char *page,
 					    size_t count)
 {
 	struct o2nm_node *node = to_o2nm_node(item);
 	struct o2nm_cluster *cluster;
-	int ret, i;
+	ssize_t ret;
+	int i;
 	struct rb_node **p, *parent;
 	unsigned int octets[4];
-	__be32 ipv4_addr = 0;
+	__be32 ipv4_addr = 0, old_ipv4_addr = 0;
+	u8 old_ipv6_addr[16] = {0};
+	int change = 0, old_ipnet_type = INVALID_IPNET_TYPE;
 
 	ret = sscanf(page, "%3u.%3u.%3u.%3u", &octets[3], &octets[2],
 		     &octets[1], &octets[0]);
@@ -290,30 +419,177 @@ static ssize_t o2nm_node_ipv4_address_store(struct config_item *item,
 	o2nm_lock_subsystem();
 	cluster = to_o2nm_cluster_from_node(node);
 	if (!cluster) {
-		o2nm_unlock_subsystem();
-		return -EINVAL;
+		ret = -EINVAL;
+		goto unlock;
 	}
 
-	ret = 0;
 	write_lock(&cluster->cl_nodes_lock);
-	if (o2nm_node_ip_tree_lookup(cluster, ipv4_addr, &p, &parent))
+	/* This means change node ip */
+	old_ipnet_type = node->nd_ipnet_type;
+
+	if (o2nm_node_ipv4_tree_lookup(cluster, ipv4_addr, &p, &parent)) {
+		write_unlock(&cluster->cl_nodes_lock);
 		ret = -EEXIST;
-	else if (test_and_set_bit(O2NM_NODE_ATTR_ADDRESS,
-			&node->nd_set_attributes))
-		ret = -EBUSY;
-	else {
-		rb_link_node(&node->nd_ip_node, parent, p);
-		rb_insert_color(&node->nd_ip_node, &cluster->cl_node_ip_tree);
+		goto unlock;
+	}
+
+	if (node->nd_ipv4_address) {
+		old_ipv4_addr = node->nd_ipv4_address;
+		rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv4_tree);
+		change = 1;
+	} else if (memcmp(node->nd_ipv6_address, old_ipv6_addr, sizeof(old_ipv6_addr))) {
+		memcpy(old_ipv6_addr, node->nd_ipv6_address, sizeof(node->nd_ipv6_address));
+		rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv6_tree);
+		change = 1;
 	}
+
+	/* get uptodate 'p' and 'parent' after rb_erase */
+	o2nm_node_ipv4_tree_lookup(cluster, ipv4_addr, &p, &parent);
+	rb_link_node(&node->nd_ip_node, parent, p);
+	rb_insert_color(&node->nd_ip_node, &cluster->cl_node_ipv4_tree);
 	write_unlock(&cluster->cl_nodes_lock);
+
+	node->nd_ipnet_type = IPV4_TYPE;
+	node->nd_ipv4_address = ipv4_addr;
+	memset(node->nd_ipv6_address, 0, sizeof(node->nd_ipv6_address));
+
+	if (change) {
+		ret = o2nm_node_change_ip_address(node);
+		if (ret) {
+			/*
+			 * rollback node's ip address when changing failed.
+			 *
+			 * 1. erase node with new ip from rbtree;
+			 * 2. set node with the old ip;
+			 * 3. reconnect or listen to all nodes with the old ip;
+			 * 4. re-insert it to rbtree with the old ip.
+			 */
+			mlog(ML_ERROR, "o2nm: node %u change ip failed.\n",
+					node->nd_num);
+			write_lock(&cluster->cl_nodes_lock);
+			rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv4_tree);
+			write_unlock(&cluster->cl_nodes_lock);
+
+			node->nd_ipnet_type = old_ipnet_type;
+			if (node->nd_ipnet_type == IPV4_TYPE)
+				node->nd_ipv4_address = old_ipv4_addr;
+			else {
+				memcpy(node->nd_ipv6_address, old_ipv6_addr,
+						sizeof(old_ipv6_addr));
+				node->nd_ipv4_address = 0;
+			}
+			o2nm_node_ip_tree_rollback(cluster, node);
+			goto unlock;
+		}
+	}
+
+	set_bit(O2NM_NODE_ATTR_ADDRESS, &node->nd_set_attributes);
+	ret = count;
+
+unlock:
 	o2nm_unlock_subsystem();
+	return ret;
+}
 
-	if (ret)
-		return ret;
+static ssize_t o2nm_node_ipv6_address_show(struct config_item *item, char *page)
+{
+	return sprintf(page, "%s\n", to_o2nm_node(item)->nd_ipv6_str);
+}
 
-	memcpy(&node->nd_ipv4_address, &ipv4_addr, sizeof(ipv4_addr));
+static ssize_t o2nm_node_ipv6_address_store(struct config_item *item,
+					    const char *page,
+					    size_t count)
+{
+	struct o2nm_node *node = to_o2nm_node(item);
+	struct o2nm_cluster *cluster;
+	ssize_t ret;
+	struct rb_node **p, *parent;
+	u8 ipv6_addr[16] = {0}, old_ipv6_addr[16] = {0};
+	int change = 0, old_ipnet_type = INVALID_IPNET_TYPE;
+	__be32 old_ipv4_addr = 0;
+
+	ret = in6_pton((const char *)page, count, ipv6_addr, -1, NULL);
+	if (ret != 1) {
+		mlog(ML_NOTICE, "in6_pton failed, ret %zd, ipv6 %s, count %zu\n",
+				ret, (const char *)page, count);
+		return -EINVAL;
+	}
 
-	return count;
+	o2nm_lock_subsystem();
+	cluster = to_o2nm_cluster_from_node(node);
+	if (!cluster) {
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	write_lock(&cluster->cl_nodes_lock);
+	/* This means change node ip */
+	old_ipnet_type = node->nd_ipnet_type;
+
+	if (o2nm_node_ipv6_tree_lookup(cluster, ipv6_addr, &p, &parent)) {
+		write_unlock(&cluster->cl_nodes_lock);
+		ret = -EEXIST;
+		goto unlock;
+	}
+
+	if (node->nd_ipv4_address) {
+		old_ipv4_addr = node->nd_ipv4_address;
+		rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv4_tree);
+		change = 1;
+	} else if (memcmp(node->nd_ipv6_address, old_ipv6_addr, sizeof(old_ipv6_addr))) {
+		memcpy(old_ipv6_addr, node->nd_ipv6_address, sizeof(node->nd_ipv6_address));
+		rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv6_tree);
+		change = 1;
+	}
+
+	/* get uptodate 'p' and 'parent' after rb_erase */
+	o2nm_node_ipv6_tree_lookup(cluster, ipv6_addr, &p, &parent);
+	rb_link_node(&node->nd_ip_node, parent, p);
+	rb_insert_color(&node->nd_ip_node, &cluster->cl_node_ipv6_tree);
+	write_unlock(&cluster->cl_nodes_lock);
+
+	node->nd_ipnet_type = IPV6_TYPE;
+	memcpy(node->nd_ipv6_address, ipv6_addr, sizeof(ipv6_addr));
+	node->nd_ipv4_address = 0;
+
+	if (change) {
+		ret = o2nm_node_change_ip_address(node);
+		if (ret) {
+			/*
+			 * rollback node's ip address when changing failed.
+			 *
+			 * 1. erase node with new ip from rbtree;
+			 * 2. set node with the old ip;
+			 * 3. reconnect or listen to all nodes with the old ip;
+			 * 4. re-insert it to rbtree with the old ip.
+			 */
+			mlog(ML_ERROR, "o2nm: node %u change ip failed.\n",
+					node->nd_num);
+			write_lock(&cluster->cl_nodes_lock);
+			rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv6_tree);
+			write_unlock(&cluster->cl_nodes_lock);
+
+			node->nd_ipnet_type = old_ipnet_type;
+			if (node->nd_ipnet_type == IPV4_TYPE) {
+				node->nd_ipv4_address = old_ipv4_addr;
+				memset(node->nd_ipv6_address, 0, sizeof(node->nd_ipv6_address));
+			} else
+				memcpy(node->nd_ipv6_address, old_ipv6_addr,
+						sizeof(old_ipv6_addr));
+			o2nm_node_ip_tree_rollback(cluster, node);
+			goto unlock;
+		}
+	}
+
+	memcpy(node->nd_ipv6_str, (char *)page, count);
+	node->nd_ipv6_str[count] = '\0';
+	set_bit(O2NM_NODE_ATTR_IPV6_ADDRESS, &node->nd_set_attributes);
+
+	ret = count;
+
+unlock:
+	o2nm_unlock_subsystem();
+	return ret;
 }
 
 static ssize_t o2nm_node_local_show(struct config_item *item, char *page)
@@ -338,36 +614,48 @@ static ssize_t o2nm_node_local_store(struct config_item *item, const char *page,
 
 	/* setting local turns on networking rx for now so we require having
 	 * set everything else first */
-	if (!test_bit(O2NM_NODE_ATTR_ADDRESS, &node->nd_set_attributes) ||
-	    !test_bit(O2NM_NODE_ATTR_NUM, &node->nd_set_attributes) ||
-	    !test_bit(O2NM_NODE_ATTR_PORT, &node->nd_set_attributes))
-		return -EINVAL; /* XXX */
+	if (!test_bit(O2NM_NODE_ATTR_PORT, &node->nd_set_attributes) ||
+	    !test_bit(O2NM_NODE_ATTR_NUM, &node->nd_set_attributes))
+		return -EINVAL;
+	if ((node->nd_ipnet_type == IPV4_TYPE) &&
+	    !test_bit(O2NM_NODE_ATTR_ADDRESS, &node->nd_set_attributes))
+		return -EINVAL;
+	else if ((node->nd_ipnet_type == IPV6_TYPE) &&
+	    !test_bit(O2NM_NODE_ATTR_IPV6_ADDRESS, &node->nd_set_attributes))
+		return -EINVAL;
 
 	o2nm_lock_subsystem();
 	cluster = to_o2nm_cluster_from_node(node);
 	if (!cluster) {
 		ret = -EINVAL;
-		goto out;
+		goto unlock;
 	}
 
 	/* the only failure case is trying to set a new local node
 	 * when a different one is already set */
 	if (tmp && tmp == cluster->cl_has_local &&
-	    cluster->cl_local_node != node->nd_num) {
+		cluster->cl_local_node != node->nd_num) {
 		ret = -EBUSY;
-		goto out;
+		goto unlock;
 	}
 
 	/* bring up the rx thread if we're setting the new local node. */
 	if (tmp && !cluster->cl_has_local) {
-		ret = o2net_start_listening(node);
+		ret = o2net_start_listening(node, false);
 		if (ret)
-			goto out;
+			goto unlock;
+
+		ret = o2net_pre_start_listening(node, false);
+		if (ret) {
+			o2net_stop_listening(node, false);
+			goto unlock;
+		}
 	}
 
 	if (!tmp && cluster->cl_has_local &&
 	    cluster->cl_local_node == node->nd_num) {
-		o2net_stop_listening(node);
+		o2net_stop_listening(node, false);
+		o2net_pre_stop_listening(node, false);
 		cluster->cl_local_node = O2NM_INVALID_NODE_NUM;
 	}
 
@@ -379,20 +667,22 @@ static ssize_t o2nm_node_local_store(struct config_item *item, const char *page,
 
 	ret = count;
 
-out:
+unlock:
 	o2nm_unlock_subsystem();
 	return ret;
 }
 
 CONFIGFS_ATTR(o2nm_node_, num);
-CONFIGFS_ATTR(o2nm_node_, ipv4_port);
+CONFIGFS_ATTR(o2nm_node_, ip_port);
 CONFIGFS_ATTR(o2nm_node_, ipv4_address);
+CONFIGFS_ATTR(o2nm_node_, ipv6_address);
 CONFIGFS_ATTR(o2nm_node_, local);
 
 static struct configfs_attribute *o2nm_node_attrs[] = {
 	&o2nm_node_attr_num,
-	&o2nm_node_attr_ipv4_port,
+	&o2nm_node_attr_ip_port,
 	&o2nm_node_attr_ipv4_address,
+	&o2nm_node_attr_ipv6_address,
 	&o2nm_node_attr_local,
 	NULL,
 };
@@ -567,16 +857,60 @@ static ssize_t o2nm_cluster_fence_method_store(
 	return -EINVAL;
 }
 
+static ssize_t o2nm_cluster_quorum_delay_threshold_show(
+		struct config_item *item, char *page)
+{
+	return sprintf(page, "%u\n", to_o2nm_cluster(item)->cl_quorum_delay_threshod);
+}
+
+static ssize_t o2nm_cluster_quorum_delay_threshold_store(
+		struct config_item *item, const char *page, size_t count)
+{
+	struct o2nm_cluster *cluster = to_o2nm_cluster(item);
+	ssize_t ret;
+	unsigned int val;
+
+	ret = o2nm_cluster_attr_write(page, count, &val);
+	if (ret > 0) {
+		if (cluster->cl_quorum_delay_threshod != val &&
+				o2net_num_connected_peers()) {
+			mlog(ML_NOTICE,
+				"o2net: cannot change quorum delay threshold after the first peer has agreed to it. %d connected peers\n",
+				o2net_num_connected_peers());
+			ret = -EINVAL;
+		} else if (val < O2NET_QUORUM_DELAY_THRESHOLD_MIN) {
+			mlog(ML_NOTICE,
+				"o2net: quorum delay threshold must be larger than %u\n",
+				O2NET_QUORUM_DELAY_THRESHOLD_MIN);
+			ret = -EINVAL;
+		} else {
+			cluster->cl_quorum_delay_threshod = val;
+		}
+	}
+
+	return ret;
+}
+
+static ssize_t o2nm_cluster_local_node_num_show(
+		struct config_item *item, char *page)
+{
+	return sprintf(page, "%u\n", o2nm_this_node());
+}
+
 CONFIGFS_ATTR(o2nm_cluster_, idle_timeout_ms);
 CONFIGFS_ATTR(o2nm_cluster_, keepalive_delay_ms);
 CONFIGFS_ATTR(o2nm_cluster_, reconnect_delay_ms);
 CONFIGFS_ATTR(o2nm_cluster_, fence_method);
+CONFIGFS_ATTR(o2nm_cluster_, quorum_delay_threshold);
+CONFIGFS_ATTR_RO(o2nm_cluster_, local_node_num);
 
 static struct configfs_attribute *o2nm_cluster_attrs[] = {
 	&o2nm_cluster_attr_idle_timeout_ms,
 	&o2nm_cluster_attr_keepalive_delay_ms,
 	&o2nm_cluster_attr_reconnect_delay_ms,
 	&o2nm_cluster_attr_fence_method,
+	&o2nm_cluster_attr_quorum_delay_threshold,
+	&o2nm_cluster_attr_local_node_num,
 	NULL,
 };
 
@@ -606,25 +940,29 @@ static void o2nm_node_group_drop_item(struct config_group *group,
 {
 	struct o2nm_node *node = to_o2nm_node(item);
 	struct o2nm_cluster *cluster = to_o2nm_cluster(group->cg_item.ci_parent);
+	u8 ipv6_addr[16] = {0};
 
 	if (cluster->cl_nodes[node->nd_num] == node) {
 		o2net_disconnect_node(node);
+		o2net_pre_shutdown_socket(node->nd_num);
 
 		if (cluster->cl_has_local &&
 		    (cluster->cl_local_node == node->nd_num)) {
 			cluster->cl_has_local = 0;
 			cluster->cl_local_node = O2NM_INVALID_NODE_NUM;
-			o2net_stop_listening(node);
+			o2net_stop_listening(node, false);
+			o2net_pre_stop_listening(node, false);
 		}
 	}
-
 	/* XXX call into net to stop this node from trading messages */
 
 	write_lock(&cluster->cl_nodes_lock);
 
 	/* XXX sloppy */
 	if (node->nd_ipv4_address)
-		rb_erase(&node->nd_ip_node, &cluster->cl_node_ip_tree);
+		rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv4_tree);
+	else if (memcmp(node->nd_ipv6_address, ipv6_addr, sizeof(ipv6_addr)))
+		rb_erase(&node->nd_ip_node, &cluster->cl_node_ipv6_tree);
 
 	/* nd_num might be 0 if the node number hasn't been set.. */
 	if (cluster->cl_nodes[node->nd_num] == node) {
@@ -711,11 +1049,13 @@ static struct config_group *o2nm_cluster_group_make_group(struct config_group *g
 	configfs_add_default_group(o2hb_group, &cluster->cl_group);
 
 	rwlock_init(&cluster->cl_nodes_lock);
-	cluster->cl_node_ip_tree = RB_ROOT;
+	cluster->cl_node_ipv4_tree = RB_ROOT;
+	cluster->cl_node_ipv6_tree = RB_ROOT;
 	cluster->cl_reconnect_delay_ms = O2NET_RECONNECT_DELAY_MS_DEFAULT;
 	cluster->cl_idle_timeout_ms    = O2NET_IDLE_TIMEOUT_MS_DEFAULT;
 	cluster->cl_keepalive_delay_ms = O2NET_KEEPALIVE_DELAY_MS_DEFAULT;
-	cluster->cl_fence_method       = O2NM_FENCE_RESET;
+	cluster->cl_fence_method       = O2NM_FENCE_INVALID;
+	cluster->cl_quorum_delay_threshod = O2NET_QUORUM_DELAY_THRESHOLD_DEFAULT;
 
 	ret = &cluster->cl_group;
 	o2nm_single_cluster = cluster;
@@ -828,7 +1168,9 @@ static int __init init_o2nm(void)
 {
 	int ret = -1;
 
-	o2hb_init();
+	ret = o2hb_init();
+	if (ret)
+		goto out;
 
 	ret = o2net_init();
 	if (ret)
diff --git a/fs/ocfs2/cluster/nodemanager.h b/fs/ocfs2/cluster/nodemanager.h
index 3e0006631cc4..f00d0c1bcd93 100644
--- a/fs/ocfs2/cluster/nodemanager.h
+++ b/fs/ocfs2/cluster/nodemanager.h
@@ -17,21 +17,33 @@
 /* This totally doesn't belong here. */
 #include <linux/configfs.h>
 #include <linux/rbtree.h>
+#include <linux/inet.h>
 
 enum o2nm_fence_method {
 	O2NM_FENCE_RESET	= 0,
 	O2NM_FENCE_PANIC,
+	O2NM_FENCE_INVALID,
 	O2NM_FENCE_METHODS,	/* Number of fence methods */
 };
 
+enum IPNET_TYPE {
+	IPV4_TYPE = 0,
+	IPV6_TYPE,
+	INVALID_IPNET_TYPE
+};
+
 struct o2nm_node {
 	spinlock_t		nd_lock;
 	struct config_item	nd_item;
+	/* to support node name change, we replace it to ip address */
 	char			nd_name[O2NM_MAX_NAME_LEN+1]; /* replace? */
-	__u8			nd_num;
+	__u16			nd_num;
 	/* only one address per node, as attributes, for now. */
+	u8			nd_ipnet_type:1;  /* 0-ipv4, 1-ipv6 */
 	__be32			nd_ipv4_address;
-	__be16			nd_ipv4_port;
+	__u8			nd_ipv6_address[16];
+	char			nd_ipv6_str[INET6_ADDRSTRLEN];
+	__be16			nd_ip_port;
 	struct rb_node		nd_ip_node;
 	/* there can be only one local node for now */
 	int			nd_local;
@@ -42,14 +54,16 @@ struct o2nm_node {
 struct o2nm_cluster {
 	struct config_group	cl_group;
 	unsigned		cl_has_local:1;
-	u8			cl_local_node;
+	u16				cl_local_node;
 	rwlock_t		cl_nodes_lock;
 	struct o2nm_node  	*cl_nodes[O2NM_MAX_NODES];
-	struct rb_root		cl_node_ip_tree;
+	struct rb_root		cl_node_ipv4_tree;
+	struct rb_root		cl_node_ipv6_tree;
 	unsigned int		cl_idle_timeout_ms;
 	unsigned int		cl_keepalive_delay_ms;
 	unsigned int		cl_reconnect_delay_ms;
 	enum o2nm_fence_method	cl_fence_method;
+	unsigned int		cl_quorum_delay_threshod;
 
 	/* this bitmap is part of a hack for disk bitmap.. will go eventually. - zab */
 	unsigned long	cl_nodes_bitmap[BITS_TO_LONGS(O2NM_MAX_NODES)];
@@ -57,11 +71,13 @@ struct o2nm_cluster {
 
 extern struct o2nm_cluster *o2nm_single_cluster;
 
-u8 o2nm_this_node(void);
+u16 o2nm_this_node(void);
+int o2nm_fence_method_is_invalid(void);
 
 int o2nm_configured_node_map(unsigned long *map, unsigned bytes);
-struct o2nm_node *o2nm_get_node_by_num(u8 node_num);
-struct o2nm_node *o2nm_get_node_by_ip(__be32 addr);
+struct o2nm_node *o2nm_get_node_by_num(u16 node_num);
+struct o2nm_node *o2nm_get_node_by_ipv4(__be32 addr);
+struct o2nm_node *o2nm_get_node_by_ipv6(u8 addr[]);
 void o2nm_node_get(struct o2nm_node *node);
 void o2nm_node_put(struct o2nm_node *node);
 
diff --git a/fs/ocfs2/cluster/ocfs2_nodemanager.h b/fs/ocfs2/cluster/ocfs2_nodemanager.h
index 21ad307419a8..8fbd56fc5320 100644
--- a/fs/ocfs2/cluster/ocfs2_nodemanager.h
+++ b/fs/ocfs2/cluster/ocfs2_nodemanager.h
@@ -15,8 +15,8 @@
 
 #define O2NM_API_VERSION	5
 
-#define O2NM_MAX_NODES		255
-#define O2NM_INVALID_NODE_NUM	255
+#define O2NM_MAX_NODES		2000
+#define O2NM_INVALID_NODE_NUM	2000
 
 /* host name, group name, cluster name all 64 bytes */
 #define O2NM_MAX_NAME_LEN        64    // __NEW_UTS_LEN
-- 
2.17.1

